ibis/common/caching.py

```diff
@@ -1,15 +1,34 @@
 from __future__ import annotations
 
+import functools
 import weakref
 from collections import Counter, defaultdict
+from collections.abc import Iterator
 from typing import Any, Callable, MutableMapping
 
-import toolz
 from bidict import bidict
 
+from ibis.common.collections import FrozenDict
 from ibis.common.exceptions import IbisError
 
 
+def memoize(func: Callable) -> Callable:
```

Comment:
Isn't this exactly what `toolz.memoize` does?

Reply:
A custom implementation of `@functools.cache` using our `frozendict`

Reply:
Ah, right! Didn't catch that at first, but I see it now!